### PR TITLE
chore: expose list of ongoing calls

### PR DESCRIPTION
**Pull request by alexandreferris**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [X] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [X] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [X] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

With now having a GlobalCallManager providing each session with its instance, we are now able to expose ongoing calls from AVS to all the way to the UI.

### Dependencies (Optional)

This is Part 2 of the epic, tests will come in a later PR.

Needs releases with:

- [X] #322 

### Attachments (Optional)

https://user-images.githubusercontent.com/5890660/160573622-02a40532-947e-4e6c-9a17-95ccc2eb3bfd.mov

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/call/CallManager.kt [hunk 1]
```diff
@@ -2,8 +2,17 @@ package com.wire.kalium.logic.feature.call
...
 expect class CallManager {
 
     suspend fun onCallingMessageReceived(message: Message, content: MessageContent.Calling)
+    val allCalls: StateFlow<List<Call>>
+}
+
+val CallManager.ongoingCalls get() = allCalls.map {
+    it.filter { call ->
+        call.status in listOf("Incoming", "Answered", "Established")
+    }
```
Comment (alexandreferris) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/call/CallManager.kt:17]: This is currently not being used, but will be in the near future as we want to properly expose each types of calls, such as:
- Ongoing calls
- Missed / Closed calls
- and so on..

**Review by typfel — CHANGES_REQUESTED** [review]
Looking good! Just some questions.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/call/Call.kt [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package com.wire.kalium.logic.feature.call
+
+data class Call(
+    val conversationId: String,
+    val status: String
```
[the remark this answers is not in the document]
Comment (ohassine) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/call/Call.kt:5]: And conversationId shouldn't be a String, it should be `ConversationId` type to support federation

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/call/GlobalCallManager.kt [hunk 3]
```diff
@@ -7,6 +7,7 @@ import com.wire.kalium.logic.data.user.UserRepository
 expect class GlobalCallManager {
 
     fun getCallManagerForClient(
+        userId: String,
```
Comment (ohassine) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/call/GlobalCallManager.kt:10]: It's a UserId not a String 🙂
As we are passing the userId, do we need the userRepository ?

File: logic/src/androidMain/kotlin/com/wire/kalium/logic/feature/call/CallManager.kt [hunk 4]
```diff
@@ -53,6 +60,28 @@ actual class CallManager(
         deferredHandle = startHandleAsync()
     }
 
+    private fun updateCallStatusById(conversationId: String, status: String) {
```
[the remark this answers is not in the document]
Comment (alexandreferris) [reply logic/src/androidMain/kotlin/com/wire/kalium/logic/feature/call/CallManager.kt:63]: Indeed! Our main focus now is making it work for Android and later on when implementing for other targets we can properly look at all the logic we have and what can and can not be extracted to the common package instead of being duplicated for all targets as you mentioned

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/call/GlobalCallManager.kt (hunk 3 above)
Comment (alexandreferris) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/call/GlobalCallManager.kt:10]: It is actually a `NonQualifiedUserId` from `network.api` package, but I have already changed it from String.

We are passing `UserId` as `NonQualifiedUserId` to use it as an initial key for holding the `CallManager` so we don't have it created multiple times for one single user.

The `userRepository` we are using to get the correct `UserId` with `QualifiedID`, thus the necessity to have both as they are exercising different things later on.

TLDR: NonQualifiedUserId for holding CallManager instances, userRepository to get user QualifiedID.

File: logic/src/androidMain/kotlin/com/wire/kalium/logic/feature/call/CallManager.kt [hunk 5]
```diff
@@ -72,19 +101,39 @@ actual class CallManager(
             incomingCallHandler = { conversationId: String, messageTime: Uint32_t, userId: String, clientId: String, isVideoCall: Boolean,
                                     shouldRing: Boolean, conversationType: Int, arg: Pointer? ->
                 kaliumLogger.i("startHandleAsync -> incomingCallHandler")
+                updateCallStatusById(
+                    conversationId = conversationId,
```
[the remark this answers is not in the document]
Comment (alexandreferris) [reply logic/src/androidMain/kotlin/com/wire/kalium/logic/feature/call/CallManager.kt:106]: It could be based on federation as far as I understood, it is now being handled in `String.toConversationId()` where it verifies if it contains a federation domain or not and it maps down to `ConversationId`

File: logic/src/jvmMain/kotlin/com/wire/kalium/logic/feature/call/GlobalCallManager.kt [hunk 6]
```diff
@@ -3,10 +3,12 @@ package com.wire.kalium.logic.feature.call
 import com.wire.kalium.logic.data.call.CallRepository
 import com.wire.kalium.logic.data.client.ClientRepository
 import com.wire.kalium.logic.data.user.UserRepository
+import com.wire.kalium.network.api.NonQualifiedUserId
 
 actual class GlobalCallManager {
 
     actual fun getCallManagerForClient(
+        userId: NonQualifiedUserId,
```
Comment (typfel) [inline logic/src/jvmMain/kotlin/com/wire/kalium/logic/feature/call/GlobalCallManager.kt:11]: You can make this a `QualifiedID` not that https://github.com/wireapp/kalium/pull/335 is merged.

Comment (typfel) [reply logic/src/jvmMain/kotlin/com/wire/kalium/logic/feature/call/GlobalCallManager.kt:11]: Don't have to be in this PR.

Comment (alexandreferris) [reply logic/src/jvmMain/kotlin/com/wire/kalium/logic/feature/call/GlobalCallManager.kt:11]: If you don't mind I will be doing this on the EPIC branch as I will need to probably solve the conflicts and thus already changing this to UserId

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/id/QualifiedId.kt [hunk 7]
```diff
@@ -1,10 +1,23 @@
...
+
 data class QualifiedID(
     val value: String,
     val domain: String
 )
 
 typealias ConversationId = QualifiedID
 
-fun QualifiedID.asString() = "$value@$domain"
+fun QualifiedID.asString() = "$value$VALUE_DOMAIN_SEPARATOR$domain"
```
Comment (ohassine) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/id/QualifiedId.kt:12]: We need to check if the domain is empty or not as well.
If we are not in federated environment, the domain will be empty
```
if(domain.isEmpty) return "$value"
else "$value$VALUE_DOMAIN_SEPARATOR$domain"
```

**Review by ohassine — CHANGES_REQUESTED** [review]
lgtm 👌 I just left one small comment

**Review by ohassine — APPROVED** [review]

File: logic/src/jvmMain/kotlin/com/wire/kalium/logic/feature/call/GlobalCallManager.kt (hunk 6 above)
Comment (alexandreferris) [reply logic/src/jvmMain/kotlin/com/wire/kalium/logic/feature/call/GlobalCallManager.kt:11]: changed it in this commit: https://github.com/wireapp/kalium/commit/833d363cef28612e61abf6184248aae6dbc24031
